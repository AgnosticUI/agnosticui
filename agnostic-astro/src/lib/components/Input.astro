---
import 'agnostic-css/src/components/input/input.css';
import 'agnostic-css/src/components/input/inputaddonitem.css';

import XElement from 'astro-xelement'
import { pluckClassesFromAttrs } from '../../shared/pluckClassesFromAttrs';

export interface Props {
  uniqueId: string;
  size?: 'small' | 'large';
  label: string;
  isInvalid?: boolean;
  invalidText?: boolean;
  helpText?: boolean;
  hasLeftAddon?: boolean;
  hasRightAddon?: boolean;
  isRounded?: boolean;
  isInlined?: boolean;
  isUnderlined?: boolean;
  isUnderlinedBackground?: boolean;
  isLabelHidden?: boolean;
  // Subset of input types e.g. this component shouldn't support button, color, or checkbox as those
  // are either covered in other components and/or a completely different UI from an "input" field
  type?: 'date' | 'datetime-local' | 'email' | 'month' | 'number' | 'password' | 'search' | 'tel' | 'text' | 'time' | 'url' | 'week';
}
// Note that as we support isInlined we need to switch from block display
// to flex-inline items-center for that case. It does couple us to some
// global utilities.
const {
  uniqueId,
  size,
  label,
  isInvalid = false,
  invalidText,
  helpText,
  hasLeftAddon = false,
  hasRightAddon = false,
  isRounded = false,
  isInlined = false,
  isUnderlined = false,
  isUnderlinedBackground = false,
  isLabelHidden = false,
  type = "text",
  ...attrs
} = Astro.props as Props;

const { input: AgInputNative, div: AgInput} = XElement;
const { additionalClassList } = pluckClassesFromAttrs(attrs);
---

<AgInput class={isInlined ? 'flex-inline items-center' : undefined}>
  <label class=`label ${isInlined ? 'label-inline' : ''}
    ${isLabelHidden ? 'screenreader-only' : ''}` for={`${uniqueId}`}>{label}</label>
  {hasLeftAddon || hasRightAddon ? (
    <div class="input-addon-container">
      <slot name="addon-left" />
      <AgInputNative
        id={`${uniqueId}`}
        name={uniqueId}
        class:list={[
          'input',
          {
            'input-rounded': isRounded ? true : false,
            'input-underlined': isUnderlined ? true : false,
            'input-underlined-bg': isUnderlinedBackground ? true : false,
            'input-small': size === 'small' ? true : false,
            'input-large': size === 'large' ? true : false,
            'input-has-left-addon': hasLeftAddon ? true : false,
            'input-has-right-addon': hasRightAddon ? true : false,
            'input-inline': isInlined ? true : false,
          },
          additionalClassList,
        ]}
        {...attrs}
      />
      <slot name="addon-right" />
    </div>
  ) : (
    <AgInputNative
      id={`${uniqueId}`}
      name={uniqueId}
      type={type}
      class:list={[
        'input',
        {
          'input-rounded': isRounded ? true : false,
          'input-underlined': isUnderlined ? true : false,
          'input-underlined-bg': isUnderlinedBackground ? true : false,
          'input-small': size === 'small' ? true : false,
          'input-large': size === 'large' ? true : false,
          'input-has-left-addon': hasLeftAddon ? true : false,
          'input-has-right-addon': hasRightAddon ? true : false,
          'input-inline': isInlined ? true : false,
          'input-error': isInvalid ? true : false,
        },
        additionalClassList,
      ]}
      {...attrs}
    />
  )}
  {isInvalid && (
    <span
      role="status"
      aria-live="polite"
      class:list={[
        {
          'field-error': !size ? true : false,
          'field-error-small': size === 'small' ? true : false,
          'field-error-large': size === 'large' ? true : false,
        }
      ]}
    >{invalidText}</span>
  )}
  {(!isInvalid && helpText) && (
    <span
      class:list={[
        {
          'field-help': !size ? true : false,
          'field-help-small': size === 'small' ? true : false,
          'field-help-large': size === 'large' ? true : false,
        }
      ]}
    >{helpText}</span>
  )}
</AgInput>
