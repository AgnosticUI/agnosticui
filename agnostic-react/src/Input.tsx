import { FC, ReactNode, HTMLAttributes } from 'react';
import styles from './input.module.css';

export interface InputProps extends HTMLAttributes<HTMLInputElement | HTMLTextAreaElement> {
  id: string;
  label: string;
  size?: 'small' | 'large';
  labelCss?: string;
  placeholder?: string;
  helpText?: string;
  invalidText?: string;
  addOnLeft?: ReactNode;
  addOnRight?: ReactNode;
  hasLeftAddon?: boolean;
  hasRightAddon?: boolean;
  css?: string;
  isInvalid?: boolean;
  isRounded?: boolean;
  isDisabled?: boolean;
  isInline?: boolean;
  isSkinned?: boolean;
  isUnderlined?: boolean;
  isUnderlinedWithBackground?: boolean;
  defaultValue?: string;
  type?: 'text' | 'password' | 'email' | 'number' | 'url' | 'tel' | 'search' | 'textarea';
  // Only useful if type is textarea
  rows?: number;
  cols?: number;
}

export const Input: FC<InputProps> = ({
  id,
  label,
  size = '',
  labelCss = '',
  placeholder = '',
  helpText = '',
  invalidText = '',
  addOnLeft = '',
  addOnRight = '',
  hasLeftAddon = false,
  hasRightAddon = false,
  css = '',
  isInvalid = false,
  isRounded = false,
  isDisabled = false,
  isInline = false,
  isSkinned = true,
  isUnderlined = false,
  isUnderlinedWithBackground = false,
  defaultValue = '',
  type = 'text',
  onChange,
  onFocus,
  onBlur,
  rows,
  cols,
  ...rest
}) => {
  const capitalize = (str: string) => str.charAt(0).toUpperCase() + str.slice(1);

  const inputClasses = [
    isSkinned ? styles.input : styles.inputBase,
    isRounded ? styles.rounded : '',
    isUnderlined ? styles.underlined : '',
    isDisabled ? styles.disabled : '',
    isInline ? styles.inputInline : '',
    hasLeftAddon ? styles.inputHasLeftAddon : '',
    hasRightAddon ? styles.inputHasRightAddon : '',
    isInvalid ? styles.invalid : '',
    isUnderlinedWithBackground ? styles.underlinedWithBackground : '',
    css ? `${css}` : '',
    size ? styles[`input${capitalize(size)}`] : '',
  ];
  const inputClassesJoined = inputClasses
    .filter((klass) => (klass ? klass.length : null))
    .join(' ');

  const labelClasses = [
    styles.label,
    isInvalid ? styles.labelError : '',
    isInline ? styles.labelInline : '',
    size ? styles[`label${capitalize(size)}`] : '',
    labelCss,
  ];
  const labelClassesJoined = labelClasses.filter((klass) => klass.length).join(' ');

  const invalidClasses = size ? styles[`fieldError${capitalize(size)}`] : styles.fieldError;
  const helpClasses = size ? styles[`fieldHelp${capitalize(size)}`] : styles.fieldHelp;
  const addonContainerClasses = styles.inputAddonContainer;

  const renderInput = () => (
    <input
      id={id}
      name={id}
      defaultValue={defaultValue}
      type={type}
      disabled={isDisabled}
      className={inputClassesJoined}
      placeholder={placeholder}
      onChange={onChange}
      onFocus={onFocus}
      onBlur={onBlur}
      {...rest}
    />
  );

  const renderAddonsOrInputs = () => {
    if (hasLeftAddon || hasRightAddon) {
      return (
        <div className={addonContainerClasses}>
          {addOnLeft && <div className={styles.leftAddon}>{addOnLeft}</div>}
          {renderInput()}
          {addOnRight && <div className={styles.rightAddon}>{addOnRight}</div>}
        </div>
      );
    }
    return renderInput();
  };

  return (
    <div className="width-full">
      <label className={labelClassesJoined} htmlFor={id}>
        {label}
      </label>
      {type === 'textarea' ? (
        <textarea
          id={id}
          name={id}
          placeholder={placeholder}
          defaultValue={defaultValue}
          className={inputClassesJoined}
          disabled={isDisabled}
          onChange={onChange}
          onFocus={onFocus}
          onBlur={onBlur}
          rows={rows}
          cols={cols}
        />
      ) : (
        renderAddonsOrInputs()
      )}
      {isInvalid && (
        <span role="status" aria-live="polite" className={invalidClasses}>
          {invalidText}
        </span>
      )}
      {helpText && <span className={helpClasses}>{helpText}</span>}
    </div>
  );
};
