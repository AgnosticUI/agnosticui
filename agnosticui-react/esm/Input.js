"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _inputModule = _interopRequireDefault(require("./input.module.css"));

var _excluded = ["addOnLeft", "addOnRight", "label", "uniqueId", "labelCss", "helpText", "invalidText", "hasLeftAddon", "hasRightAddon", "isInvalid", "isRounded", "inputCss", "isDisabled", "isInline", "isSkinned", "isUnderlinedWithBackground", "isUnderlined", "size", "defaultValue", "type", "onChange", "onFocus", "onBlur", "placeholder"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Input = function Input(_ref) {
  var addOnLeft = _ref.addOnLeft,
      addOnRight = _ref.addOnRight,
      label = _ref.label,
      uniqueId = _ref.uniqueId,
      labelCss = _ref.labelCss,
      helpText = _ref.helpText,
      invalidText = _ref.invalidText,
      hasLeftAddon = _ref.hasLeftAddon,
      hasRightAddon = _ref.hasRightAddon,
      isInvalid = _ref.isInvalid,
      isRounded = _ref.isRounded,
      inputCss = _ref.inputCss,
      isDisabled = _ref.isDisabled,
      isInline = _ref.isInline,
      isSkinned = _ref.isSkinned,
      isUnderlinedWithBackground = _ref.isUnderlinedWithBackground,
      isUnderlined = _ref.isUnderlined,
      size = _ref.size,
      defaultValue = _ref.defaultValue,
      type = _ref.type,
      onChange = _ref.onChange,
      onFocus = _ref.onFocus,
      onBlur = _ref.onBlur,
      placeholder = _ref.placeholder,
      rest = _objectWithoutProperties(_ref, _excluded);

  var capitalize = function capitalize(str) {
    return str.charAt(0).toUpperCase() + str.slice(1);
  };

  var inputClasses = [isSkinned ? _inputModule["default"].input : _inputModule["default"].inputBase, isRounded ? _inputModule["default"].rounded : '', isUnderlined ? _inputModule["default"].underlined : '', isDisabled ? _inputModule["default"].disabled : '', isInline ? _inputModule["default"].inputInline : '', hasLeftAddon ? _inputModule["default"].inputHasLeftAddon : '', hasRightAddon ? _inputModule["default"].inputHasRightAddon : '', isInvalid ? _inputModule["default"].invalid : '', isUnderlinedWithBackground ? _inputModule["default"].underlinedWithBackground : '', inputCss ? "".concat(inputCss) : '', size ? _inputModule["default"]["input".concat(capitalize(size))] : ''];
  inputClasses = inputClasses.filter(function (klass) {
    return klass ? klass.length : null;
  });
  inputClasses = inputClasses.join(' ');
  var labelClasses = [_inputModule["default"].label, isInvalid ? _inputModule["default"].labelError : '', isInline ? _inputModule["default"].labelInline : '', size ? _inputModule["default"]["label".concat(capitalize(size))] : '', labelCss ? labelCss : ''];
  labelClasses = labelClasses.filter(function (klass) {
    return klass.length;
  });
  labelClasses = labelClasses.join(' ');
  var invalidClasses = size ? _inputModule["default"]["fieldError".concat(capitalize(size))] : _inputModule["default"].fieldError;
  var helpClasses = size ? _inputModule["default"]["fieldHelp".concat(capitalize(size))] : _inputModule["default"].fieldHelp;
  var addonContainerClasses = _inputModule["default"].inputAddonContainer;

  var renderInput = function renderInput() {
    return /*#__PURE__*/_react["default"].createElement("input", _extends({
      id: uniqueId,
      name: uniqueId,
      defaultValue: defaultValue,
      type: type,
      disabled: isDisabled,
      className: inputClasses,
      placeholder: placeholder,
      onChange: onChange,
      onFocus: onFocus,
      onBlur: onBlur
    }, rest));
  };

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: "width-full"
  }, /*#__PURE__*/_react["default"].createElement("label", {
    className: labelClasses,
    htmlFor: uniqueId
  }, label), type == 'textarea' ? /*#__PURE__*/_react["default"].createElement("textarea", _extends({
    id: uniqueId,
    name: uniqueId,
    placeholder: placeholder,
    defaultValue: defaultValue,
    className: inputClasses,
    disabled: isDisabled
  }, rest)) : hasLeftAddon || hasRightAddon ? /*#__PURE__*/_react["default"].createElement("div", {
    className: addonContainerClasses
  }, addOnLeft && /*#__PURE__*/_react["default"].createElement("div", {
    className: _inputModule["default"].leftAddon
  }, addOnLeft), renderInput(), addOnRight && /*#__PURE__*/_react["default"].createElement("div", {
    className: _inputModule["default"].rightAddon
  }, addOnRight)) : renderInput(), isInvalid && /*#__PURE__*/_react["default"].createElement("span", {
    role: "status",
    "aria-live": "polite",
    className: invalidClasses
  }, invalidText), helpText && /*#__PURE__*/_react["default"].createElement("span", {
    className: helpClasses
  }, helpText));
};

Input.propTypes = {
  size: _propTypes["default"].oneOf(['small', 'large']),
  label: _propTypes["default"].string.isRequired,
  uniqueId: _propTypes["default"].string.isRequired,
  labelCss: _propTypes["default"].string,
  placeholder: _propTypes["default"].string,
  helpText: _propTypes["default"].string,
  invalidText: _propTypes["default"].string,
  addOnLeft: _propTypes["default"].node,
  addOnRight: _propTypes["default"].node,
  hasLeftAddon: _propTypes["default"].bool,
  hasRightAddon: _propTypes["default"].bool,
  isInvalid: _propTypes["default"].bool,
  isRounded: _propTypes["default"].bool,
  inputCss: _propTypes["default"].string,
  isDisabled: _propTypes["default"].bool,
  isInline: _propTypes["default"].bool,
  isSkinned: _propTypes["default"].bool,
  isUnderlinedWithBackground: _propTypes["default"].bool,
  isUnderlined: _propTypes["default"].bool,
  defaultValue: _propTypes["default"].string,
  type: _propTypes["default"].oneOf(['text', 'password', 'email', 'number', 'url', 'tel', 'search', 'textarea']),
  onChange: _propTypes["default"].func,
  onFocus: _propTypes["default"].func,
  onBlur: _propTypes["default"].func,
  // Useful for valid html attribute e.g. if you have type textarea you may want to use rows or cols
  rest: _propTypes["default"].any
};
Input.defaultProps = {
  placeholder: undefined,
  labelCss: undefined,
  helpText: undefined,
  invalidText: undefined,
  addOnLeft: undefined,
  addOnRight: undefined,
  hasLeftAddon: false,
  hasRightAddon: false,
  isInvalid: false,
  isRounded: false,
  inputCss: undefined,
  isSkinned: true,
  isDisabled: false,
  isInline: false,
  isUnderlined: false,
  isUnderlinedWithBackground: false,
  size: undefined,
  defaultValue: '',
  type: 'text',
  onChange: undefined,
  onFocus: undefined,
  onBlur: undefined
};
var _default = Input;
exports["default"] = _default;