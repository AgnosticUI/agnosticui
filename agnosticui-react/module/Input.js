var _excluded = ["addOnLeft", "addOnRight", "label", "uniqueId", "labelCss", "helpText", "invalidText", "hasLeftAddon", "hasRightAddon", "isInvalid", "isRounded", "inputCss", "isDisabled", "isInline", "isSkinned", "isUnderlinedWithBackground", "isUnderlined", "size", "defaultValue", "type", "onChange", "onFocus", "onBlur", "placeholder"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import styles from './input.module.css';

var Input = function Input(_ref) {
  var addOnLeft = _ref.addOnLeft,
      addOnRight = _ref.addOnRight,
      label = _ref.label,
      uniqueId = _ref.uniqueId,
      labelCss = _ref.labelCss,
      helpText = _ref.helpText,
      invalidText = _ref.invalidText,
      hasLeftAddon = _ref.hasLeftAddon,
      hasRightAddon = _ref.hasRightAddon,
      isInvalid = _ref.isInvalid,
      isRounded = _ref.isRounded,
      inputCss = _ref.inputCss,
      isDisabled = _ref.isDisabled,
      isInline = _ref.isInline,
      isSkinned = _ref.isSkinned,
      isUnderlinedWithBackground = _ref.isUnderlinedWithBackground,
      isUnderlined = _ref.isUnderlined,
      size = _ref.size,
      defaultValue = _ref.defaultValue,
      type = _ref.type,
      onChange = _ref.onChange,
      onFocus = _ref.onFocus,
      onBlur = _ref.onBlur,
      placeholder = _ref.placeholder,
      rest = _objectWithoutProperties(_ref, _excluded);

  var capitalize = function capitalize(str) {
    return str.charAt(0).toUpperCase() + str.slice(1);
  };

  var inputClasses = [isSkinned ? styles.input : styles.inputBase, isRounded ? styles.rounded : '', isUnderlined ? styles.underlined : '', isDisabled ? styles.disabled : '', isInline ? styles.inputInline : '', hasLeftAddon ? styles.inputHasLeftAddon : '', hasRightAddon ? styles.inputHasRightAddon : '', isInvalid ? styles.invalid : '', isUnderlinedWithBackground ? styles.underlinedWithBackground : '', inputCss ? "".concat(inputCss) : '', size ? styles["input".concat(capitalize(size))] : ''];
  inputClasses = inputClasses.filter(function (klass) {
    return klass ? klass.length : null;
  });
  inputClasses = inputClasses.join(' ');
  var labelClasses = [styles.label, isInvalid ? styles.labelError : '', isInline ? styles.labelInline : '', size ? styles["label".concat(capitalize(size))] : '', labelCss ? labelCss : ''];
  labelClasses = labelClasses.filter(function (klass) {
    return klass.length;
  });
  labelClasses = labelClasses.join(' ');
  var invalidClasses = size ? styles["fieldError".concat(capitalize(size))] : styles.fieldError;
  var helpClasses = size ? styles["fieldHelp".concat(capitalize(size))] : styles.fieldHelp;
  var addonContainerClasses = styles.inputAddonContainer;

  var renderInput = function renderInput() {
    return /*#__PURE__*/React.createElement("input", _extends({
      id: uniqueId,
      name: uniqueId,
      defaultValue: defaultValue,
      type: type,
      disabled: isDisabled,
      className: inputClasses,
      placeholder: placeholder,
      onChange: onChange,
      onFocus: onFocus,
      onBlur: onBlur
    }, rest));
  };

  return /*#__PURE__*/React.createElement("div", {
    className: "width-full"
  }, /*#__PURE__*/React.createElement("label", {
    className: labelClasses,
    htmlFor: uniqueId
  }, label), type == 'textarea' ? /*#__PURE__*/React.createElement("textarea", _extends({
    id: uniqueId,
    name: uniqueId,
    placeholder: placeholder,
    defaultValue: defaultValue,
    className: inputClasses,
    disabled: isDisabled
  }, rest)) : hasLeftAddon || hasRightAddon ? /*#__PURE__*/React.createElement("div", {
    className: addonContainerClasses
  }, addOnLeft && /*#__PURE__*/React.createElement("div", {
    className: styles.leftAddon
  }, addOnLeft), renderInput(), addOnRight && /*#__PURE__*/React.createElement("div", {
    className: styles.rightAddon
  }, addOnRight)) : renderInput(), isInvalid && /*#__PURE__*/React.createElement("span", {
    role: "status",
    "aria-live": "polite",
    className: invalidClasses
  }, invalidText), helpText && /*#__PURE__*/React.createElement("span", {
    className: helpClasses
  }, helpText));
};

Input.propTypes = {
  size: PropTypes.oneOf(['small', 'large']),
  label: PropTypes.string.isRequired,
  uniqueId: PropTypes.string.isRequired,
  labelCss: PropTypes.string,
  placeholder: PropTypes.string,
  helpText: PropTypes.string,
  invalidText: PropTypes.string,
  addOnLeft: PropTypes.node,
  addOnRight: PropTypes.node,
  hasLeftAddon: PropTypes.bool,
  hasRightAddon: PropTypes.bool,
  isInvalid: PropTypes.bool,
  isRounded: PropTypes.bool,
  inputCss: PropTypes.string,
  isDisabled: PropTypes.bool,
  isInline: PropTypes.bool,
  isSkinned: PropTypes.bool,
  isUnderlinedWithBackground: PropTypes.bool,
  isUnderlined: PropTypes.bool,
  defaultValue: PropTypes.string,
  type: PropTypes.oneOf(['text', 'password', 'email', 'number', 'url', 'tel', 'search', 'textarea']),
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  // Useful for valid html attribute e.g. if you have type textarea you may want to use rows or cols
  rest: PropTypes.any
};
Input.defaultProps = {
  placeholder: undefined,
  labelCss: undefined,
  helpText: undefined,
  invalidText: undefined,
  addOnLeft: undefined,
  addOnRight: undefined,
  hasLeftAddon: false,
  hasRightAddon: false,
  isInvalid: false,
  isRounded: false,
  inputCss: undefined,
  isSkinned: true,
  isDisabled: false,
  isInline: false,
  isUnderlined: false,
  isUnderlinedWithBackground: false,
  size: undefined,
  defaultValue: '',
  type: 'text',
  onChange: undefined,
  onFocus: undefined,
  onBlur: undefined
};
export default Input;