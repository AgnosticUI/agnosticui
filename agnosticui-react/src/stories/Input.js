import React from 'react';
import PropTypes from 'prop-types';
import styles from './input.module.css';

const Input = ({
  label,
  uniqueId,
  labelCss,
  helpText,
  invalidText,
  hasLeftAddon,
  hasRightAddon,
  isInvalid,
  isRounded,
  inputCss,
  isDisabled,
  isInline,
  isSkinned,
  isUnderlinedWithBackground,
  isUnderlined,
  size,
  defaultValue,
  type,
  onChange,
  onFocus,
  onBlur,
  placeholder,
  rest,
}) => {
  const capitalize = str => str.charAt(0).toUpperCase() + str.slice(1);
  let inputClasses = [
    isSkinned ? styles.input : styles.inputBase,
    isRounded ? styles.rounded : '',
    isUnderlined ? styles.underlined : '',
    isDisabled ? styles.disabled : "",
    isInline ? styles.inputInline : "",
    hasLeftAddon ? styles.leftAddon : '',
    hasRightAddon ? styles.rightAddon : '',
    isInvalid ? styles.invalid : '',
    isUnderlinedWithBackground ? styles.underlinedWithBackground : '',
    inputCss ? `${inputCss}` : '',
    size ? styles[`input${capitalize(size)}`] : '',
  ];
  inputClasses = inputClasses.filter((klass) => klass.length);
  inputClasses = inputClasses.join(" ");

  let labelClasses = [
    styles.label,
    isInvalid ? styles.labelError : "",
    isInline ? styles.labelInline : "",
    size ? styles[`label${capitalize(size)}`] : "",
    labelCss ? labelCss : "",
  ];
  labelClasses = labelClasses.filter((klass) => klass.length);
  labelClasses = labelClasses.join(" ");

  const invalidClasses = size ? styles[`fieldError${capitalize(size)}`] : styles.fieldError;
  const helpClasses = size ? styles[`fieldHelp${capitalize(size)}`] : styles.fieldHelp;
  const addonContainerClasses = styles.inputAddonContainer;

  const renderInput = () => {
    return (
      <input
        id={uniqueId}
        name={uniqueId}
        defaultValue={defaultValue}
        type={type}
        disabled={isDisabled}
        className={inputClasses}
        placeholder={placeholder}
        onChange={onChange}
        onFocus={onFocus}
        onBlur={onBlur}
      />
    )
  }
  return (
    <div className="width-full">
      <label className={labelClasses} htmlFor={uniqueId}>{label}</label>
      {type == 'textarea' && (
        <textarea
          id={uniqueId}
          name={uniqueId}
          placeholder={placeholder}
          defaultValue={defaultValue}
          className={inputClasses}
          disabled={isDisabled}
          {...rest}
        />
      )}
      {hasLeftAddon || hasRightAddon ? (
        <div className={addonContainerClasses}>
          {addOnLeft && (
            <div>{addOnLeft}</div>
          )}
          {renderInput()}
          {addOnRight && (
            <div>{addOnRight}</div>
          )}
        </div>
      ) : (
          <>
            { renderInput()}
          </>
        )
      }

      {isInvalid && (
        <span className={invalidClasses}>{invalidText}</span>
      )}
      {helpText && (
        <span className={helpClasses}>{helpText}</span>
      )}
    </div>
  );
};

Input.propTypes = {
  size: PropTypes.oneOf(['small', 'large']),
  label: PropTypes.string.isRequired,
  uniqueId: PropTypes.string.isRequired,
  labelCss: PropTypes.string,
  placeholder: PropTypes.string,
  helpText: PropTypes.string,
  invalidText: PropTypes.string,
  hasLeftAddon: PropTypes.bool,
  hasRightAddon: PropTypes.bool,
  isInvalid: PropTypes.bool,
  isRounded: PropTypes.bool,
  inputCss: PropTypes.string,
  isDisabled: PropTypes.bool,
  isInline: PropTypes.bool,
  isSkinned: PropTypes.bool,
  isUnderlinedWithBackground: PropTypes.bool,
  isUnderlined: PropTypes.bool,
  defaultValue: PropTypes.string,
  type: PropTypes.oneOf(['text', 'password', 'email', 'number', 'url', 'tel', 'search', 'textarea']),
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
}

Input.defaultProps = {
  placeholder: undefined,
  labelCss: undefined,
  helpText: undefined,
  invalidText: undefined,
  hasLeftAddon: false,
  hasRightAddon: false,
  isInvalid: false,
  isRounded: false,
  inputCss: undefined,
  isSkinned: true,
  isDisabled: false,
  isInline: false,
  isUnderlined: false,
  isUnderlinedWithBackground: false,
  size: undefined,
  defaultValue: '',
  type: 'text',
  onChange: undefined,
  onFocus: undefined,
  onBlur: undefined,
}

export default Input;
