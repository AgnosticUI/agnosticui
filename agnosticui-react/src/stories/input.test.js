import React from 'react';
import '@testing-library/jest-dom/extend-expect'
import { render, fireEvent } from '@testing-library/react'
import "regenerator-runtime/runtime.js";
import Input from './Input.js'

test('onChange', async () => {
  const changeHandler = jest.fn()
  const { getByRole } = render(<Input label="label is require prop haha" uniqueId="myUniqId" onChange={changeHandler} />);
  const input = getByRole('textbox')
  const mockTypingEvent = {
    target: {
      value: "changed-value"
    }
  };
  await fireEvent.change(input, mockTypingEvent)
  expect(changeHandler).toHaveBeenCalledTimes(1)
})


test('onFocus', async () => {
  const focusHandler = jest.fn()
  const { getByRole } = render(<Input label="label is require prop haha" uniqueId="myUniqId" onFocus={focusHandler} />);
  const input = getByRole('textbox')
  const mockTypingEvent = {
    target: {
      value: "focus-value"
    }
  };
  await fireEvent.focus(input, mockTypingEvent)
  expect(focusHandler).toHaveBeenCalledTimes(1)
})

test('onBlur', async () => {
  const blurHandler = jest.fn()
  const { getByRole } = render(<Input label="label is require prop haha" uniqueId="myUniqId" onBlur={blurHandler} />);
  const input = getByRole('textbox')
  const mockTypingEvent = {
    target: {
      value: "blur-value"
    }
  };
  await fireEvent.blur(input, mockTypingEvent)
  expect(blurHandler).toHaveBeenCalledTimes(1)
})
