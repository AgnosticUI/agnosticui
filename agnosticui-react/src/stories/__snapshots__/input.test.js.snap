// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="width-full"
      >
        <label
          class="label"
          for="myUniqId"
        >
          label is require prop haha
        </label>
        <input
          class="input"
          id="myUniqId"
          name="myUniqId"
          type="text"
          value=""
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="width-full"
    >
      <label
        class="label"
        for="myUniqId"
      >
        label is require prop haha
      </label>
      <input
        class="input"
        id="myUniqId"
        name="myUniqId"
        type="text"
        value=""
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
